internal/redactor/redactor.go

```diff
@@ -0,0 +1,425 @@
+// Package redactor provides an efficient configurable string redactor.
+package redactor
+
+import (
+	"errors"
+	"io"
+	"path"
+	"sync"
+
+	"github.com/buildkite/agent/v3/bootstrap/shell"
+)
+
+// RedactLengthMin is the shortest string length that will be considered a
+// potential secret by the environment redactor. e.g. if the redactor is
+// configured to filter out environment variables matching *_TOKEN, and
+// API_TOKEN is set to "none", this minimum length will prevent the word "none"
+// from being redacted from useful log output.
+const RedactLengthMin = 6
+
+// Redactor is a straightforward secret redactor.
+//
+// The algorithm is intended to be easier to maintain than certain
+// high-performance multi-string replacement algorithms, and also geared towards
+// ensuring secrets don't escape (for instance, by matching overlaps), at the
+// expense of ultimate efficiency.
+type Redactor struct {
+	// Replacement string (e.g. "[REDACTED]")
+	subst []byte
+
+	// Secrets to redact (looking for these needles in the haystack),
+	// organised by first byte.
+	// Why first byte? Because looking up needles by the first byte is a lot
+	// faster than _filtering_ all the needles by first byte.
+	needlesByFirstByte [256][]string
+
+	// For synchronising writes. Each write can touch everything below.
+	mu sync.Mutex
+
+	// Redacted output written to this writer.
+	dst io.Writer
+
+	// Intermediate buffer to account for partially-written non-secrets.
+	// (i.e. we began redacting in case we're in the middle of a secret, but
+	// we might not be).
+	buf []byte
+
+	// Current redaction states - if we have begun redacting a potential secret
+	// there will be at least one of these.
+	// nextStates is the next set of states. To avoid creating millions of new
+	// slices, Write alternates between these two.
+	states, nextStates []state
```

Comment:
can you go into a bit more detail (here in the PR, rather than in code) as to why we have the two state arrays and we swap through them? i'm not 100% sure i understand how/why this helps us avoid creating millions of new slices

Reply:
🤔 given that `states` is a pretty vague name, (and if i'm reading what these slices actually do correctly), would it be slightly clearer to name these `partialMatches` (and `nextPartialMatches` i guess? `partialMatches1` and partialmatches2`?) and `redact` something like `completeMatches`?

Reply:
I added more comment. I also went with `partialMatches`, `nextMatches`, and `completedMatches`, thoughts?

Reply:
fantastic - love it. much better than `partialMatches1` and `partialMatches2` 😅